Source/WebKit/UIProcess/ios/WKContentView.mm

```diff
@@ -921,38 +930,45 @@ - (NSUInteger)_wk_pageCountForPrintFormatter:(_WKWebViewPrintFormatter *)printFo
             pageCount = _page->computePagesForPrintingiOS(frameID, printInfo);
     }
 
-    ASSERT(!_pdfPrintCompletionSemaphore);
-    if (isPrintingOnBackgroundThread)
-        _pdfPrintCompletionSemaphore = makeUnique<BinarySemaphore>();
-
-    // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
-    _pdfPrintCallbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf = retainPtr(self)](RefPtr<WebCore::SharedBuffer>&& pdfData) {
-        if (!pdfData || pdfData->isEmpty())
-            retainedSelf->_printedDocument = nullptr;
-        else {
-            auto data = pdfData->createCFData();
-            auto dataProvider = adoptCF(CGDataProviderCreateWithCFData(data.get()));
-            retainedSelf->_printedDocument = adoptCF(CGPDFDocumentCreateWithProvider(dataProvider.get()));
-        }
+    {
+        Locker locker { _pdfPrintCompletionLock };
 
-        if (auto pdfPrintCompletionSemaphore = std::exchange(retainedSelf->_pdfPrintCompletionSemaphore, nullptr))
-            pdfPrintCompletionSemaphore->signal();
+        ASSERT(!_pdfPrintCompletionCondition);
+        if (isPrintingOnBackgroundThread)
+            _pdfPrintCompletionCondition = makeUnique<Condition>();
+    }
 
-        retainedSelf->_pdfPrintCallbackID = 0;
+    ensureOnMainRunLoop([frameID, printInfo, pageCount, self, retainedSelf = retainPtr(self)] {
+        // Begin generating the PDF in expectation of a (eventual) request for the drawn data.
+        _pdfPrintCallbackID = _page->drawToPDFiOS(frameID, printInfo, pageCount, [retainedSelf](RefPtr<WebCore::SharedBuffer>&& pdfData) {
+            if (!pdfData || pdfData->isEmpty())
+                retainedSelf->_printedDocument = nullptr;
+            else {
+                auto data = pdfData->createCFData();
+                auto dataProvider = adoptCF(CGDataProviderCreateWithCFData(data.get()));
+                retainedSelf->_printedDocument = adoptCF(CGPDFDocumentCreateWithProvider(dataProvider.get()));
+            }
+
+            [retainedSelf _resetPrintingState];
+        });
     });
 
     return pageCount;
 }
 
 - (BOOL)_waitForDrawToPDFCallbackIfNeeded
 {
-    if (auto callbackID = std::exchange(_pdfPrintCallbackID, 0)) {
-        if (_pdfPrintCompletionSemaphore)
-            _pdfPrintCompletionSemaphore->wait();
-        else if (!_page->process().connection()->waitForAsyncCallbackAndDispatchImmediately<Messages::WebPage::DrawToPDFiOS>(callbackID, Seconds::infinity()))
-            return false;
+    if (isMainRunLoop()) {
```

Comment:
What if the first print is done on a background thread, but then a second print is done on the main thread?  Shouldn't we still wait for the `_pdfPrintCompletionCondition`?

Reply:
Also going to test this scenario.